### PR TITLE
Prevent overlapping arrays for AES-CBC

### PR DESCRIPTION
**Pull request by ashbm5**
Signed-off-by: Alon <alonsh@ca.ibm.com>

### CODE REVIEW

**Comment by ashbm5** [comment]
Fix for https://github.com/eclipse-openj9/openj9/issues/14597

fyi @keithc-ca @pshipton

File: closed/src/java.base/share/classes/com/sun/crypto/provider/NativeCipherBlockChaining.java [hunk 1]
```diff
@@ -260,6 +260,21 @@ int encrypt(byte[] plain, int plainOffset, int plainLen,
             throw new ProviderException("Internal error in input buffering");
         }
 
+        /**
+         * OpenSSL doesn't support overlapping buffers.
+         * If plaintext and ciphertext byte arrays are overlapping, make a copy of plain
+         */
+        byte[] copyOfInput = null;
+        if (plain == cipher) {
+            if ((plainOffset >= cipherOffset) ||
```
Comment (pshipton) [inline closed/src/java.base/share/classes/com/sun/crypto/provider/NativeCipherBlockChaining.java:269]: Seems this part should also check the length to ensure there is an overlap.

Comment (ashbm5) [reply closed/src/java.base/share/classes/com/sun/crypto/provider/NativeCipherBlockChaining.java:269]: the returned ciphertext's length is undetermined and may span all remaining space starting from cipherOffset.

File: closed/src/java.base/share/classes/com/sun/crypto/provider/NativeCipherBlockChaining.java [hunk 2]
```diff
@@ -260,6 +260,21 @@ int encrypt(byte[] plain, int plainOffset, int plainLen,
...
+                ((plainOffset < cipherOffset) && (cipherOffset < (plainOffset + plainLen)))) {
+                copyOfInput = new byte[plainLen];
+                System.arraycopy(plain,  plainOffset, copyOfInput, 0, plainLen);
+                plain = copyOfInput;
+                plainOffset = 0;
+            }
+        }
```
Comment (keithc-ca) [inline closed/src/java.base/share/classes/com/sun/crypto/provider/NativeCipherBlockChaining.java:276]: The declaration of `copyOfInput` should move into the `if` statement.

Is it really worth checking the offsets (`encryptFinal()` does not)? If so:
* the overlap test includes a redundant check (`a || (!a && b))` is just `a || b`)
* the overlap test is overly strong when `plainOffset >= cipherOffset` (there's no overlap if `plainOffset + plainLen >= cipherOffset`)

File: closed/src/java.base/share/classes/com/sun/crypto/provider/NativeCipherBlockChaining.java [hunk 3]
```diff
@@ -260,6 +260,21 @@ int encrypt(byte[] plain, int plainOffset, int plainLen,
             throw new ProviderException("Internal error in input buffering");
         }
 
+        /**
+         * OpenSSL doesn't support overlapping buffers.
+         * If plaintext and ciphertext byte arrays are overlapping, make a copy of plain
```
Comment (keithc-ca) [inline closed/src/java.base/share/classes/com/sun/crypto/provider/NativeCipherBlockChaining.java:265]: Please finish sentences with a period.

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: closed/src/java.base/share/classes/com/sun/crypto/provider/NativeCipherBlockChaining.java (hunk 1 above)
Comment (pshipton) [reply closed/src/java.base/share/classes/com/sun/crypto/provider/NativeCipherBlockChaining.java:269]: Pls add a comment to explain this.

Comment (ashbm5) [reply closed/src/java.base/share/classes/com/sun/crypto/provider/NativeCipherBlockChaining.java:269]: this java code invokes the OpenSSL routine EVP_CipherUpdate which is a generic function that can handle multiple modes of encryption/decryption. Tracking the output buffer size and ensuring it is sufficient is encryption mode dependent and is up to the code that uses OpenSSL. In our case, the code in [CipherCore.java ](https://github.com/ibmruntimes/openj9-openjdk-jdk/blob/42e7b85c8a915406a79fdef29d7180e5a06f16b9/src/java.base/share/classes/com/sun/crypto/provider/CipherCore.java#L578) manages the needed sizes and may buffer input if the output buffer is not sufficient. Turns out for AES-CBC mode update operation, the input/output buffers sizes should be identical. I will modify the check to reflect that.

File: closed/src/java.base/share/classes/com/sun/crypto/provider/NativeCipherBlockChaining.java (hunk 3 above)
Comment (ashbm5) [reply closed/src/java.base/share/classes/com/sun/crypto/provider/NativeCipherBlockChaining.java]: updated

File: closed/src/java.base/share/classes/com/sun/crypto/provider/NativeCipherBlockChaining.java (hunk 2 above)
Comment (ashbm5) [reply closed/src/java.base/share/classes/com/sun/crypto/provider/NativeCipherBlockChaining.java]: updated.

File: closed/src/java.base/share/classes/com/sun/crypto/provider/NativeCipherBlockChaining.java [hunk 4]
```diff
@@ -260,6 +260,18 @@ int encrypt(byte[] plain, int plainOffset, int plainLen,
...
         }
 
+        /**
+         * OpenSSL doesn't support overlapping buffers.
+         * If plaintext and ciphertext byte arrays are overlapping, make a copy of plain.
+         */
+        if (plain == cipher) {
+            byte[] copyOfInput = null;
+            copyOfInput = new byte[plainLen];
+            System.arraycopy(plain,  plainOffset, copyOfInput, 0, plainLen);
```
Comment (keithc-ca) [inline closed/src/java.base/share/classes/com/sun/crypto/provider/NativeCipherBlockChaining.java:270]: Please join the declaration and the assignment:
```
    byte[] copyOfInput = new byte[plainLen];
```
Please also remove the extra space after the first comma.

**Comment by keithc-ca** [comment]
I'd appreciate an answer to my earlier question:
> What prevents the native code from writing beyond the end of the array?

For example, I could find no check equivalent to the one in the non-native implementation: https://github.com/ibmruntimes/openj9-openjdk-jdk/blob/openj9/src/java.base/share/classes/com/sun/crypto/provider/CipherBlockChaining.java#L151

File: closed/src/java.base/share/classes/com/sun/crypto/provider/NativeCipherBlockChaining.java (hunk 4 above)
Comment (ashbm5) [reply closed/src/java.base/share/classes/com/sun/crypto/provider/NativeCipherBlockChaining.java]: done

File: closed/src/java.base/share/classes/com/sun/crypto/provider/NativeCipherBlockChaining.java [hunk 5]
```diff
@@ -260,6 +260,17 @@ int encrypt(byte[] plain, int plainOffset, int plainLen,
             throw new ProviderException("Internal error in input buffering");
         }
 
+        /**
+         * OpenSSL doesn't support overlapping buffers.
+         * If plaintext and ciphertext byte arrays are overlapping, make a copy of plain.
```
Comment (keithc-ca) [inline closed/src/java.base/share/classes/com/sun/crypto/provider/NativeCipherBlockChaining.java:265]: This comment is out-of-date now.

Comment (ashbm5) [reply closed/src/java.base/share/classes/com/sun/crypto/provider/NativeCipherBlockChaining.java]: updated

Comment (keithc-ca) [reply closed/src/java.base/share/classes/com/sun/crypto/provider/NativeCipherBlockChaining.java]: Thanks. In the future, please don't use javadoc-style comments in places like this.

**Review by keithc-ca — APPROVED** [review]

**Comment by keithc-ca** [comment]
@ashbm5 Why s your commit signed-off-by alonsh?
```
Signed-off-by: alonsh <alonsh@ca.ibm.com>
```
Why did Alon not create this pull request himself?

Also, I'd still like an answer about the (apparently missing) bounds-checking; see https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/406#issuecomment-1054531262.

**Comment by ashbm5** [comment]
I will change sign off, about the checks: this is the check to ensure output array is sufficiently large:

https://github.com/ibmruntimes/openj9-openjdk-jdk/blob/42e7b85c8a915406a79fdef29d7180e5a06f16b9/src/java.base/share/classes/com/sun/crypto/provider/CipherCore.java#L637

those bounds checks we don't have seem to be related to the hotspot JIT compiler replacing the methods at runtime, see https://hg.openjdk.java.net/jdk7u/jdk7u/jdk/rev/d15ed96035d0

**Comment by pshipton** [comment]
@ashbm5 you don't have to of course, but it would be handy if you added your name to your github profile.

**Comment by pshipton** [comment]
jenkins test sanity.openjdk alinux64 jdknext

**Comment by keithc-ca** [comment]
jenkins signed off by check

**Comment by pshipton** [comment]
jenkins signed off by check

**Comment by pshipton** [comment]
jenkins copyright check

**Comment by pshipton** [comment]
jenkins line endings check

**Comment by pshipton** [comment]
jenkins test sanity.openjdk alinux64 jdknext

**Comment by pshipton** [comment]
com/sun/crypto/provider/Cipher/AES/TestCopySafe ran and passed as expected. The failures are unrelated to his PR and have open issues.

**Review by pshipton — APPROVED** [review]

**Comment by pshipton** [comment]
@ashbm5 I assume this should be backported to the previous releases?

**Comment by ashbm5** [comment]
@pshipton yes @WilburZjh will port to jdk8/11/17/18 once this PR is merged

**Comment by pshipton** [comment]
Hmm, I though I merged it. Trying again.

**Pull request merged**